### PR TITLE
Derive profile name from display name in `ProfileService.CreateProfile`

### LINKED ISSUE #2943: Derive `Name` from `DisplayName`
### Please describe the enhancement

`Name` is required to be in a specific format, whereas `DisplayName` is more freeform. When creating something new (such as a profile) using the API, the caller needs to have a display name, and will possibly derive the name by doing certain transformation (regexp replace, trim, etc) on the display name.

This means that clients of the API might implement this logic in different ways, and the logic is duplicated within the ecosystem, increasing the overall complexity in the ecosystem.

### Solution Proposal

The logic to derive Name from DisplayName could be moved to all POST endpoints in the API. This would mean that the clients don't need to directly supply the Name value when creating a new object from scratch, and this logic could be removed from all client code.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Acceptance Criteria

_No response_

### PR DESCRIPTION
**Pull request by gajananan**
# Summary

Currently, clients of the API are required to derive the profile name from the display name using various transformations. This results in duplicated logic across the ecosystem, increasing complexity and potential inconsistencies.

This commit scopes the logic to derive the profile name from the display name to the ProfileService.CreateProfile endpoint. By centralizing this logic, we reduce redundancy and simplify client code. Clients of the API are no longer need to implement logic to deriive the profile name when creating a new profile.

Fixes #2943 

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: internal/profiles/util.go [hunk 1]
```diff
@@ -299,6 +301,59 @@ func MergeDatabaseGetByNameIntoProfiles(ppl []db.GetProfileByProjectAndNameRow)
...
+
+		profileName = CleanDisplayName(displayName)
+
+	}
+
+	return profileName
+}
+
+// The profile name should be derived from the profile display name given the following logic
+func CleanDisplayName(displayName string) string {
```
Comment (jhrozek) [inline internal/profiles/util.go:331]: this function probably doesn't have to be exported

File: internal/profiles/util.go [hunk 2]
```diff
@@ -299,6 +301,59 @@ func MergeDatabaseGetByNameIntoProfiles(ppl []db.GetProfileByProjectAndNameRow)
...
+
+	// Trim leading and trailing whitespace
+	displayName = strings.TrimSpace(displayName)
+
+	// Remove non-alphanumeric characters
+	re := regexp.MustCompile(`[^a-zA-Z0-9\s]`)
```
Comment (jhrozek) [inline internal/profiles/util.go:337]: if the regexes are not going to change, then it could probably be a global but non exported variable

**Review by jhrozek — COMMENTED** [review]
Thank you for your contribution @gajananan !

I'm going to tag @blkt who was recently discussing the semantics of display names and might have better opinions on that than I. In the meantime I just put some general comments into the PR and I'm going to let the CI pipeline run.

**Review by eleftherias — COMMENTED** [review]
This is looking great so far @gajananan.
I know you had some questions about his requirement:

> If the derived profile name already exists in the same project, then we append a -1 to the end of the derived profile name (my_profile-1).
If the new name also exists in the project, then we increment the appending digit by 1 until we find a name that doesn't already exist. (my_profile-2, my_profile-3 etc)

To fulfil this, you will have to add an extra call to fetch all the profiles from the the current project, and then pass their names to `DeriveProfileNameFromDisplayName`.

Once the "clean" profile name is derived, you'll need to check it against all the existing names.

Some example inputs and outputs:
`DeriveProfileNameFromDisplayName("My profile", ["other_profile", "custom_profile")` -> `"my_profile"`

`DeriveProfileNameFromDisplayName("My profile", ["other_profile", "my_profile")` -> `"my_profile-1"`

`DeriveProfileNameFromDisplayName("My profile", ["other_profile", "my_profile", "my_profile-1")` -> `"my_profile-2"`

Feel free to reach out if this is unclear or if you have any questions.

File: internal/profiles/util.go (hunk 1 above)
Comment (gajananan) [reply internal/profiles/util.go]: Thank you, @jhrozek. Fixed now.

File: internal/profiles/util.go (hunk 2 above)
Comment (gajananan) [reply internal/profiles/util.go]: Thank you, @jhrozek. Fixed now

File: internal/profiles/util.go [hunk 3]
```diff
@@ -299,6 +301,59 @@ func MergeDatabaseGetByNameIntoProfiles(ppl []db.GetProfileByProjectAndNameRow)
...
+
+	} else if displayName != "" && name == "" {
+		// when a display name is provided, but no name
+		// then the profile name is created and saved based on the profile display name
+
+		profileName = CleanDisplayName(displayName)
+
+	}
+
+	return profileName
```
[the remark this answers is not in the document]
Comment (gajananan) [reply internal/profiles/util.go]: thank you, @eleftherias. Fixed now.

**Comment by gajananan** [comment]
> This is looking great so far @gajananan. I know you had some questions about his requirement:
> 
> > If the derived profile name already exists in the same project, then we append a -1 to the end of the derived profile name (my_profile-1).
> > If the new name also exists in the project, then we increment the appending digit by 1 until we find a name that doesn't already exist. (my_profile-2, my_profile-3 etc)
> 
> To fulfil this, you will have to add an extra call to fetch all the profiles from the the current project, and then pass their names to `DeriveProfileNameFromDisplayName`.
> 
> Once the "clean" profile name is derived, you'll need to check it against all the existing names.
> 
> Some example inputs and outputs: `DeriveProfileNameFromDisplayName("My profile", ["other_profile", "custom_profile")` -> `"my_profile"`
> 
> `DeriveProfileNameFromDisplayName("My profile", ["other_profile", "my_profile")` -> `"my_profile-1"`
> 
> `DeriveProfileNameFromDisplayName("My profile", ["other_profile", "my_profile", "my_profile-1")` -> `"my_profile-2"`
> 
> Feel free to reach out if this is unclear or if you have any questions.

I have updated the PR to implement this.  Please let me know if the approach needs to change.

File: internal/profiles/util_test.go [hunk 4]
```diff
@@ -634,3 +634,126 @@ func TestFilterRulesForType(t *testing.T) {
...
+			name: "Derived profile name does exist in the current project",
+			profile: &minderv1.Profile{
+				Name:        "",
+				DisplayName: "My profile",
+			},
+			existingProfileNames: []string{"other_profile", "my_profile"},
+			expected:             "my_profile-1",
+		},
+		{
+			name: "Derived profile name does exist in the current project",
```
Comment (eleftherias) [inline internal/profiles/util_test.go:737]: There are 2 tests with the same name "Derived profile name does exist in the current project", which will make debugging difficult if one fails.

File: internal/profiles/util.go [hunk 5]
```diff
@@ -299,6 +305,61 @@ func MergeDatabaseGetByNameIntoProfiles(ppl []db.GetProfileByProjectAndNameRow)
...
+	}
+	// when both a display name and a profile name are provided
+	// then the profile name from the incoming request is used as the profile name
+
+	derivedName := name
+	counter := 1
+
+	// check if the current project already has a profile with that name, then add a counter
+	for strings.Contains(strings.Join(existingProfileNames, " "), derivedName) {
+		derivedName = fmt.Sprintf("%s-%d", name, counter)
```
[the remark this answers is not in the document]
Comment (gajananan) [reply internal/profiles/util.go:332]: Thank you, I fixed the logic to handle the edge cases mentioned as well as added unit tests for them.  @eleftherias

**Review by eleftherias — APPROVED** [review]

**Comment by eleftherias** [comment]
Thank you for you contribution @gajananan!

**Comment by gajananan** [comment]
> Thank you for you contribution @gajananan!

Thank you @eleftherias for clarifying the requirements and the follow up feedback.
